File: yukiko/ruby/45_kaiten_sushi.rb
class Solve
  def initialize
    @n = gets.to_i
    @v_list = gets.chomp.split.map(&:to_i)
    @dp = Array.new(@n, 0)
  end

  def run
    return call(0)
  end
  
  private

  def call(idx)
    return 0 if idx >= @n
    return @dp[idx] if @dp[idx] > 0
    return @dp[idx] = [call(idx+1), call(idx+2) + @v_list[idx]].max
  end
end

puts Solve.new.run

File: yukiko/python/45_kaiten_sushi.py
N = int(input())
sushi = [int(i) for i in input().split()]
dp = [0]*N
dp[N-1] = sushi[N-1]
dp[N-2] = max(sushi[N-2:])

for i in reversed(range(N-2)):
    dp[i] = max(sushi[i] + dp[i+2], dp[i+1])

print(dp[0])
